[MRG] Dynamic Topic Models in python.

bhargavvader opened the pull request:
A python translation from C of Dynamic Topic Models, as described by Blei at al in this [paper](http://icml.cc/2016/awards/dtm.pdf), originally written in C [here](https://github.com/blei-lab/dtm). Dynamic Topic Models are used to topic model time-series tagged documents such that the topics evolve over each time-slice. This is useful in finding out how certain key words in topics change over time and in finding document similarity between documents far apart in time but contextually similar.

This was my Google Summer of Code 2016 project.

TODO:

    1) Include Document Influence Model (DIM) mode. Most of the infrastructure for this is in place.
    2) See if LdaPost can be replaced by LdaModel completely without breaking anything.
        - in particular, a lot of DIM depends on LdaPost being in place.
    3) Heavy lifting going on in the sslm class - efforts can be made to cythonise mathematical methods.
        - in particular, update_obs and the optimization takes a lot time.
    4) Try and make it distributed, especially around the E and M step.
    5) Get rid of all C/C++ coding styles if left behind.

tmylk commented:
Could you please update it to be more up to date with your private fork?

bhargavvader commented:
@tmylk all the methods are more or less done. I still have to see `scipy` implementations of the `gsl` methods and replace them, as well as add tests for the entire module.

bhargavvader commented:
Note: as of now, the classes and methods are not well arranged, and there are a few mock classes (which will be removed) to help me with testing. Once all the testing is done and a crude DTM is fit, I will rearrange the structure of the DTM code.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1055 @@
+import sys
+
+# this is a mock LDA class to help with testing until this is figured out
+class mockLDA(utils.SaveLoad):
```
should the mock class
 be in test folder rather than in prod file?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1055 @@
+    def __init__(self, num_terms=None, max_nterms=None, length=None, num_docs=None, corpuses=None, corpus=None):
+        self.num_terms = num_terms
+        self.max_nterms = max_nterms
+        self.length = len(corpuses)
```
what is  the difference between nterms and length?
Please add a doc-string

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1055 @@
+        self.renormalized_influence = renormalized_influence
+        self.influence_sum_lgl = influence_sum_lgl
+
+class sslm(utils.SaveLoad):
```
docstring please

bhargavvader commented:
Cleaned up code, added docstrings so that it is more reviewable now. Made changes to many core classes so most of the tests would be failing; will fix them and further clean tomorrow.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+        if self.time_slice is not None:
+            self.num_sequences = len(time_slice)
+
+        # need to still figure out way to get max_nterms    
```
Mark TODO with `TODO:`

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    num_sequences is the number of sequences, i.e number of time-slices.
+    num_docs is the number of documents present.
+    time_slice is a list or numpy array which the user must provide which contains the number of documents in each time-slice.
+    corpus is any iterable gensim corpus.
```
can you change names to `vocab_len`, `max_doc_len`, `num_time_slices`, `corpus_len` ?

bhargavvader replied:
done

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+class LdaSeqModel(utils.SaveLoad):
+    """
+    Class which contains information of our whole DTM model.
+    Topic chains contains for each topic a 'state space language model' object which in turn has information about each topic
```
add more description on what sslm is.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    """
+    obs values contain the doc - topic ratios
+    e_log_prob contains topic - word ratios
+    mean, fwd_mean contains the mean values to be used for inference for each word for a time_slice
```
what is the difference between mean and fwd_mean?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+            self.id2word = utils.dict_from_corpus(corpus)
+            self.num_terms = len(self.id2word)
+        elif len(self.id2word) > 0:
+            self.num_terms = 1 + max(self.id2word.keys())
```
why is this different from L49 above?

bhargavvader replied:
changed it to be the same.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+        self.obs = numpy.array(numpy.split(numpy.zeros(num_sequences * num_terms), num_terms))
+        self.e_log_prob = numpy.array(numpy.split(numpy.zeros(num_sequences * num_terms), num_terms))
+        self.mean = numpy.array(numpy.split(numpy.zeros((num_sequences + 1) * num_terms), num_terms))
+        self.fwd_mean = numpy.array(numpy.split(numpy.zeros((num_sequences + 1) * num_terms), num_terms))
```
why do you need `+1` sometimes? looks a bit hacky

bhargavvader replied:
The +1 is used as an extra time-slice to help with calculations.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    for i in range(0, num_terms):
+        for j in range(0, num_sequences):
+            m = sslm.mean[i][j + 1]
+            v = sslm.variance[i][j + 1]
```
would it be easier to `zip` `sslm.mean` and `sslm.variance`?

bhargavvader replied:
I guess I could... would it speed things up though? Or for readability?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+        w = sslm.obs_variance / (fwd_variance[t - 1] + chain_variance + sslm.obs_variance)
+        fwd_mean[t] = w * fwd_mean[t - 1] + (1 - w) * obs[t - 1]
+        if fwd_mean[t] is None:
+            # error message
```
add the error msg

bhargavvader replied:
This was just done at that point to replicate the C code; that error can't happen so I've removed that bit.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+            w = chain_variance / (fwd_variance[t] + chain_variance)
+        mean[t] = w * fwd_mean[t] + (1 - w) * mean[t + 1]
+        if mean[t] is None:
+            # error message
```
same as prev comment

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    W = sslm.num_terms
+    T = sslm.num_sequences
+    for t in range(0, T):
+        for w in range(0, W):
```
just use enumerate

bhargavvader replied:
In come cases such as in compute_post_mean where I iterate to a num_sequence + 1th slice for mean and fwd_mean, I would need it. The places I don't I could use enumerate. Is that ok?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+            w = 0
+        variance[t] = (w * (variance[t + 1] - chain_variance)) + ((1 - w) * fwd_variance[t])
+
+    sslm.variance[word] = variance
```
can it just return the new variance and fwd_variance? Functions with side effects are hard to read. Same for `compute_post_mean` below

bhargavvader replied:
Yeah will make those changes.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+
+    # compute post variance
+    for w in range(0, W):
+       compute_post_variance(w, sslm, sslm.chain_variance)
```
better to show explicitly the changes to sslm, no side effects. see comment in `compute_post_variance` above

bhargavvader replied:
yes

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+            v = sslm.variance[i][j + 1]
+            val = numpy.exp(m + v/2)
+            sslm.zeta[j] = sslm.zeta[j] + val  
+    return
```
please return new `zeta` instead of assigning it here

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    final_iters_flag = 0 
+    last_iter = 0
+
+    # this is a flag/input do something about it
```
add a TODO to this comment

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+        print ("E Step")
+
+        # writing to file
+        em_log.write(str(bound) + "\t" + str(convergence))
```
what is `em_log`? do you mean to use logger?

bhargavvader replied:
The C code writes these to disk. I'll just give the option of saving the model to disk and won't need this.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+
+
+        if ((bound - old_bound) < 0):
+            if (LDA_INFERENCE_MAX_ITER == 1):
```
what is the purpose of this?
also it can be less lines if it is a `dict` or just multiply by 2

bhargavvader replied:
will use dict.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    lda_post = Lda_Post(max_nterms=seq_corpus.max_nterms, num_topics=K, lda=lda)
+
+
+    model = "DTM"
```
add a comment

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+
+    model = "DTM"
+    if model == "DTM":
+        inferDTMseq(K, ldaseq, seq_corpus, topic_suffstats, gammas, lhoods, iter_, last_iter, lda, lda_post, bound)
```
make immutable

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+    converged = 0
+    iter_ = 0
+    LDA_INFERENCE_CONVERGED = 1e-8
+    LDA_INFERENCE_MAX_ITER = 25
```
move all constants to the top

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+        for k in range(0, K):
+            lda_post.gamma[k] += phi_row[k] * count
+
+    return
```
just return gamma instead of changing it

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1121 @@
+            args = sslm, w_counts, totals, mean_deriv_mtx, w, deriv
+            obs = sslm.obs[w]
+            step_size = 0.01
+            tol = 1e-3
```
make constants in CAPS

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+
+
+class LdaSeqModel(utils.SaveLoad):
+    """
```
Add how to use similar to https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/ldamodel.py#L200

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+        self.word = word
+        self.count = count
+        self.total = total
+
```
add endclass markers for easier reading

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    """
+    Updates the Zeta Variational Parameter.
+    Zeta is described in the appendix and is equal to sum (exp(mean[word] + Variance[word] / 2)), over every time-slice.
+    Zeta is obtained as a term which we obtain after solving the equation to maximize the lower bound as a function of the variational parameters (mean and variance).
```
Rephrase please

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    num_time_slices = sslm.num_time_slices
+    sslm.zeta.fill(0)
+
+    zeta = sslm.zeta
```
unnecessary alias

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    fwd_variance = sslm.fwd_variance[word]
+
+    # forward pass. Set initial variance very high
+    fwd_variance[0] = chain_variance * 1000
```
no magic numbers. make a named constant

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    This is done based on the original Blei-LDA paper, where:
+    log_phi := beta * exp(Ψ(gamma)), over every topic for every word.
+    
+    TODO: incorporate lee-sueng trick used in **Lee, Seung: Algorithms for non-negative matrix factorization, NIPS 2001**.
```
nice TODO!

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    W = sslm.vocab_len
+    T = sslm.num_time_slices
+    for t in range(0, T):
+        for w in range(0, W):
```
can this be vectorized?

bhargavvader replied:
It can't be for this right? At best I can do for `w`. And by vectorise here do you mean exactly? I still need values of `w` and `t` to access my `mean` values right?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+
+    # setting variational observations to transformed counts
+    for t in range(0, T):
+        sslm.obs[:,t] = log_norm_counts
```
vectorize instead of `for`

bhargavvader replied:
Even if I use `numpy.vstack` I would have to iterate - because I am going column by column. Do you know any other way I can fill all columns of a matrix with the same values? @piskvorky any help here?

bhargavvader replied:
@dsquareindia suggested using `numpy.repeat` to create a temp array and assign this to `sslm.obs` - this will work fine, apart from the fact that I might have a large temp array before assigning it to `sslm.obs`. Should that be a problem?

dsquareindia replied:
you don't need to create a temp array. I was saying:
```python
>>> a = np.arange(9).reshape(3,3)
>>> a
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> # If you want to repeat the second column all over:
... a = np.repeat(a[:,1].reshape(-1, 1), 5, axis=1)
>>> a
array([[1, 1, 1, 1, 1],
       [4, 4, 4, 4, 4],
       [7, 7, 7, 7, 7]])
>>> 
```
Although I'm not sure how big of a memory advantage this is.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+
+    ldaseq.alphas = alpha
+    for k in range(0, ldaseq.num_topics):
+
```
remove line

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+def lda_seq_infer(ldaseq, seq_corpus, topic_suffstats, gammas, lhoods, iter_, last_iter):
+
+    """
+    set up lda model to use for inferDTMseq
```
expand docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+def inferDTMseq(K, ldaseq, seq_corpus, topic_suffstats, gammas, lhoods, iter_, last_iter, lda, lda_post, bound):
+
+    """
+    compute the likelihood of a sequential corpus under an LDA seq model. return the likelihood bound.
```
expand docstring describing params

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    Initialize variational posterior, does not return anything.
+    """
+    K = lda_post.lda.num_topics
+    N = lda_post.doc.nterms
```
K,N unnecessary aliases

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+
+    sslm.zeta = update_zeta(sslm)
+
+    return sslm.obs
```
why not `sslm.zeta` as well? or entire `sslm`?

bhargavvader replied:
Yeah, I'll return `sslm` and just assign `zeta` and `obs` where it was called.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+            w = sslm.chain_variance / (fwd_variance[t] + sslm.chain_variance)
+        deriv[t] = w * deriv[t] + (1 - w) * deriv[t + 1]
+
+    return
```
return the affected `deriv`

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    
+    lda = ldamodel.LdaModel(num_topics=K, alpha=ldaseq.alphas, id2word=seq_corpus.id2word)
+    lda.topics = numpy.array(numpy.split(numpy.zeros(W * K), W))
+    lda_post = Lda_Post(max_doc_len=seq_corpus.max_doc_len, num_topics=K, lda=lda)
```
weird casing in `Lda_Post`

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1222 @@
+    if model == "DTM":
+        inferDTMseq(K, ldaseq, seq_corpus, topic_suffstats, gammas, lhoods, iter_, last_iter, lda, lda_post, bound)
+    elif model == "DIM":
+        InfluenceTotalFixed(ldaseq, seq_corpus);
```
we don't use camel casing in python. use underscore

tmylk commented:
Could you please upload the concise tests as well?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1159 @@
+
+    def __getitem__(self, doc):
+        """
+        TODO: To mimic the __getitem__ in ldamodel. This method is a work in progress.
```
what is left to do here?

bhargavvader replied:
Finished the method now, it needed some checking.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+        Method to initialize State Space Language Model, topic wise.
+        """
+        self.alphas = alpha
+        for k in range(0, self.num_topics):
```
`enumerate` is more concise

bhargavvader replied:
Yeah, will iterate over chains.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+            bound += topic_bound
+
+            if ((bound - old_bound) < 0):
+                if LDA_INFERENCE_MAX_ITER < 10:
```
why magic numbers?

bhargavvader replied:
Will make user flag.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+
+            if convergence < LDASQE_EM_THRESHOLD:
+
+                LDA_INFERENCE_MAX_ITER = 500
```
CAPITALs mean that it is a constant. If it changes then it should be a variable

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+        lhood = 0
+        lhood_term = 0
+
+        for k in range(0, self.num_topics):
```
just iterate over topic_chains

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+        self.obs_variance = obs_variance
+        self.chain_variance = chain_variance
+
+        # compute post variance
```
you mean `post mean and variance`? 
A single for loop for both is enough

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+        totals = numpy.zeros(counts.shape[1])
+
+        for w in range(0, W):
+            self.variance[w], self.fwd_variance[w] = self.compute_post_variance(w, self.chain_variance)
```
`zip*` and list comprehension is easier to read

bhargavvader replied:
Will do.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+        set up the LDA model topic-word values with that of ldaseq.
+        """
+        num_topics = self.num_topics
+        for k in range(0, num_topics):
```
no need for an alias `num_topics`

piskvorky commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1096 @@
+            if ((bound - old_bound) < 0):
+                if LDA_INFERENCE_MAX_ITER < 10:
+                    LDA_INFERENCE_MAX_ITER *= 2
+                print ("Bound went down, increasing iterations to", LDA_INFERENCE_MAX_ITER)
```
All these `print` statements will go away, right?

@bhargavvader it is a better practice to use logging from the start. Just use DEBUG level if you want the messages to go away in "normal" mode (logging threshold set at INFO+).

tmylk replied:
Yes, we discussed it during our calls.

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+
+    1) Include DIM mode. Most of the infrastructure for this is in place.
+    2) See if LdaPost can be replaces by LdaModel completely without breakign anything.
+    3) Heavy lifting going on in the sslm class - efforts can be made to cythonise mathematical methods.
```
how much time is spent there according to line profiler? is it really a perf bottleneck?

bhargavvader replied:
`fit_sslm` takes up majority of the time with `update_obs` being particularly slow.

tmylk commented on gensim/test/test_ldaseqmodel.py:
```diff
@@ -0,0 +1,46 @@
+        ['river','water','mud','tree'],['money','transaction','bank','finance'],
+        ['bank','borrow','money'], ['bank','finance'], ['finance','money','sell','bank'],['borrow','sell'],['bank','loan','sell']]
+        # initializing using own LDA sufficient statistics so that we get same results each time.
+        sstats = numpy.loadtxt(datapath('sstats_test'))
```
Please rename to sstats_test.txt so the format is clear

tmylk commented on gensim/test/test_ldaseqmodel.py:
```diff
@@ -0,0 +1,46 @@
+        self.ldaseq = ldaseqmodel.LdaSeqModel(corpus = corpus , id2word= dictionary, num_topics=2, time_slice=[10, 10, 11], initialize='own', sstats=sstats)
+
+    def testTopicWord(self):
+
```
extra line

tmylk commented on gensim/test/test_ldaseqmodel.py:
```diff
@@ -0,0 +1,46 @@
+        corpus = [dictionary.doc2bow(text) for text in texts]
+        self.ldaseq = ldaseqmodel.LdaSeqModel(corpus = corpus , id2word= dictionary, num_topics=2, time_slice=[10, 10, 11], initialize='own', sstats=sstats)
+
+    def testTopicWord(self):
```
what is tested here? Please add a docstring

tmylk commented on gensim/test/test_ldaseqmodel.py:
```diff
@@ -0,0 +1,46 @@
+        self.assertAlmostEqual(topics[0][0][0], expected_topic_word[0][0], places=2)
+        self.assertEqual(topics[0][0][1], expected_topic_word[0][1])
+
+    def testDocTopic(self):
```
what is tested here? Please add a docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+
+            if convergence < LDASQE_EM_THRESHOLD:
+
+                lda_inference_max_iter = 500
```
magic number?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+        num_time_slices = self.num_time_slices
+        self.zeta.fill(0)
+        for j in range(0, num_time_slices):
+            self.zeta[j] = numpy.sum(numpy.exp(self.mean[:, j + 1] + self.variance[:, j + 1] / 2))
```
why not just iterate over zeta?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+        """
+        vocab_len = self.vocab_len
+        num_time_slices = self.num_time_slices
+        self.zeta.fill(0)
```
is it needed in python?

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+    def fit_sslm(self, counts):
+
+        """
+        Fit variational distribution.
```
expand the docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+
+    def sslm_counts_init(self, obs_variance, chain_variance, sstats):
+        """
+        Initialize State Space Language Model with LDA sufficient statistics.
```
expand the docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+        """
+        Fucntion to perform optimization of obs.
+        This is by far the slowest function in the whole algorithm.
+        Replacing or improving the performance of this would greatly speed things up.
```
expand the docstring with description of arguments

tmylk replied:
make it a TODO

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+
+    def update_lda_seq_ss(self, time, doc, topic_suffstats):
+        """
+        Update lda sequence sufficient statistics from an lda posterior.
```
expand the docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -0,0 +1,1100 @@
+def f_obs(x, *args):
+
+    """
+    Function which we are optimising for minimizing obs
```
expand the docstring

tmylk commented on gensim/models/ldaseqmodel.py:
```diff
@@ -921,7 +926,7 @@ def compute_lda_lhood(self):
         # sigma_d = 0 
 
-        lhood = math.lgamma(numpy.sum(self.lda.alpha)) - math.lgamma(gamma_sum)
+        lhood = gammaln(numpy.sum(self.lda.alpha)) - gammaln(gamma_sum)
```
why this change?

bhargavvader replied:
`math.lgamma` doesn't exist in python 2.6. `scipy.special.gammln`does the exact same thing and is there.

bhargavvader commented:
@tmylk , @piskvorky , I've added a tutorial notebook and tests to accompany the code. 
Could you do a review so I can start working on changes?

tmylk commented:
@piskvorky I think this is ready for your review. (Though I still have some comments about the code and notebook to discuss with Bhargav today.)

piskvorky commented:
OK, I'll have a look, probably this weekend. Thanks @bhargavvader !

One thing I see immediately is that this PR needs a better description. When people come here from the CHANGELOG/web/google/blog etc, they shouldn't be greeted with `"This is a very, very rough draft"`.

bhargavvader commented:
@piskvorky , haha, yes. I've changed the title and description with a brief explanation and TODO for further works.

piskvorky commented:
@tmylk I see a lot of TODOs in the description -- was this really meant to be merged?

bhargavvader commented:
@piskvorky , the TODOs in the description are more of further works - things like improving performance and making it distributed. I've changed it from TODO to `Future Works` to reflect this.
I am opening another PR to work on the tutorial notebook, and on incorporating suggestions.
